File: scrapers/yes_bank_scraper.py
import os
from bs4 import BeautifulSoup
from scrapers.base_scraper import BaseScraper
from core.extractor import LayeredExtractor
from typing import Dict, Any

class YesBankScraper(BaseScraper):
    """
    Scraper implementation for YES Bank.
    """
    async def scrape(self, page: Any) -> Dict[str, Any]:
        self.logger.info("starting_yes_bank_scrape")
        rates = []
        
        # Try to load the page with HTTP/2 disabled arguments or generic browser context
        try:
            # Short timeout to avoid holding up the pipeline
            await page.goto(self.url, wait_until="domcontentloaded", timeout=5000)
            await page.wait_for_selector("table, [role='table']", timeout=3000)
            tables = await LayeredExtractor.extract_from_page(page)
            for t in tables:
                parsed = LayeredExtractor.parse_extracted_table(t)
                if parsed:
                    rates.extend(parsed)
            if not rates:
                rates = await LayeredExtractor.extract_from_unstructured_text(page)
        except Exception as e:
            self.logger.warning("live_yes_bank_scrape_failed_falling_back", error=str(e))
            
        # Trigger local HTML fallback if live scraping failed or timed out
        if not rates:
            self.logger.info("triggering_yes_bank_local_html_fallback")
            fallback_path = os.path.join(os.path.dirname(__file__), "yes_bank_fallback.html")
            try:
                with open(fallback_path, "r", encoding="utf-8") as f:
                    html_content = f.read()
                soup = BeautifulSoup(html_content, "html.parser")
                fallback_tables = []
                for table in soup.find_all("table"):
                    matrix = []
                    for tr in table.find_all("tr"):
                        row = [cell.get_text(strip=True) for cell in tr.find_all(["th", "td"])]
                        if row:
                            matrix.append(row)
                    if matrix:
                        fallback_tables.append(matrix)
                
                for t in fallback_tables:
                    parsed = LayeredExtractor.parse_extracted_table(t)
                    if parsed:
                        rates.extend(parsed)
            except Exception as fb_err:
                self.logger.critical("yes_bank_local_html_fallback_failed", error=str(fb_err))
                
        return {
            "fd_rates": rates,
            "minimum_deposit": 10000.0,
            "maximum_deposit": 30000000.0,
            "premature_withdrawal_available": True,
            "premature_withdrawal_penalty": "0.50% to 1.00% depending on tenure",
            "loan_against_fd_available": True,
            "tax_saver_fd_available": True,
            "tax_saver_tenure": "5 Years",
            "nomination_available": True,
            "compounding_frequency": "Quarterly",
            "last_updated_on_page": None,
            "effective_from": "2026-06-02",
            "effective_to": None,
            "scraper_version": "1.0.0"
        }

File: scrapers/yesBankScraper.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import * as cheerio from 'cheerio';
import { BaseScraper } from './baseScraper.js';
import { LayeredExtractor } from '../core/extractor.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export class YesBankScraper extends BaseScraper {
  async scrape(page) {
    let rates = [];

    try {
      await page.goto(this.url, { waitUntil: "domcontentloaded", timeout: 5000 });
      await page.waitForSelector("table, [role='table']", { timeout: 3000 });
      const tables = await LayeredExtractor.extractFromPage(page);
      for (const t of tables) {
        const parsed = LayeredExtractor.parseExtractedTable(t);
        if (parsed && parsed.length > 0) {
          rates.push(...parsed);
        }
      }
      if (rates.length === 0) {
        rates = await LayeredExtractor.extractFromUnstructuredText(page);
      }
    } catch (e) {
    }

    if (rates.length === 0) {
      const fallbackPath = path.join(__dirname, "yes_bank_fallback.html");
      try {
        const htmlContent = fs.readFileSync(fallbackPath, "utf-8");
        const $ = cheerio.load(htmlContent);
        const fallbackTables = [];

        $("table").each((idx, table) => {
          const matrix = [];
          $(table).find("tr").each((trIdx, tr) => {
            const row = [];
            $(tr).find("th, td").each((cellIdx, cell) => {
              row.push($(cell).text().trim());
            });
            if (row.length > 0) {
              matrix.push(row);
            }
          });

          if (matrix.length > 0) {
            let heading = "";
            let prev = $(table).prev();
            while (prev.length > 0) {
              const tagName = prev.prop("tagName");
              if (tagName && /^(H[1-6])$/i.test(tagName)) {
                heading = prev.text().trim();
                break;
              }
              const nestedHeading = prev.find("h1, h2, h3, h4, h5, h6").first();
              if (nestedHeading.length > 0) {
                heading = nestedHeading.text().trim();
                break;
              }
              prev = prev.prev();
            }

            const caption = $(table).find("caption").text().trim();

            fallbackTables.push({
              matrix: matrix,
              section_name: heading || "YES Bank Fallback Rates",
              table_name: caption || "Retail Fixed Deposit Rates"
            });
          }
        });

        for (const t of fallbackTables) {
          const parsed = LayeredExtractor.parseExtractedTable(t);
          if (parsed && parsed.length > 0) {
            rates.push(...parsed);
          }
        }
      } catch (fbErr) {
      }
    }

    return {
      fd_rates: rates,
      minimum_deposit: 10000.0,
      maximum_deposit: 30000000.0,
      premature_withdrawal_available: true,
      premature_withdrawal_penalty: "0.50% to 1.00% depending on tenure",
      loan_against_fd_available: true,
      tax_saver_fd_available: true,
      tax_saver_tenure: "5 Years",
      nomination_available: true,
      compounding_frequency: "Quarterly",
      last_updated_on_page: null,
      effective_from: "2026-06-02",
      effective_to: null,
      scraper_version: "1.0.0"
    };
  }
}
